World.py

```diff
@@ -1009,107 +1034,59 @@ def set_goals(self):
+        if isinstance(region_name, Region):
+            return region_name
+        if (region := self._region_cache.get(region_name, None)) is not None:
+            return region
 
-    def get_entrance(self, entrance):
-        if isinstance(entrance, Entrance):
+        for region in self.regions:
+            self._region_cache[region.name] = region
+        if (region := self._region_cache.get(region_name, None)) is not None:
+            return region
+        raise KeyError('No such region %s' % region_name)
+
+    def get_entrance(self, entrance_name: str | Entrance) -> Entrance:
+        if isinstance(entrance_name, Entrance):
+            return entrance_name
+        if (entrance := self._entrance_cache.get(entrance_name, None)) is not None:
             return entrance
-        try:
-            return self._entrance_cache[entrance]
-        except KeyError:
-            for region in self.regions:
-                for exit in region.exits:
-                    if exit.name == entrance:
-                        self._entrance_cache[entrance] = exit
-                        return exit
-            raise KeyError('No such entrance %s' % entrance)
 
+        for region in self.regions:
+            for entrance in region.exits:
+                self._entrance_cache[entrance.name] = entrance
+        if (entrance := self._entrance_cache.get(entrance_name, None)) is not None:
+            return entrance
+        raise KeyError('No such entrance %s' % entrance_name)
 
-    def get_location(self, location):
-        if isinstance(location, Location):
+    def get_location(self, location_name: str | Location) -> Location:
+        if isinstance(location_name, Location):
+            return location_name
+        if (location := self._location_cache.get(location_name, None)) is not None:
             return location
-        try:
-            return self._location_cache[location]
-        except KeyError:
-            for region in self.regions:
-                for r_location in region.locations:
-                    if r_location.name == location:
-                        self._location_cache[location] = r_location
-                        return r_location
-        raise KeyError('No such location %s' % location)
-
-
-    def get_items(self):
-        return [loc.item for loc in self.get_filled_locations()] + self.itempool
 
+        for region in self.regions:
+            for location in region.locations:
+                self._location_cache[location.name] = location
+        if (location := self._location_cache.get(location_name, None)) is not None:
+            return location
+        raise KeyError('No such location %s' % location_name)
 
-    def get_itempool_with_dungeon_items(self):
-        return self.get_restricted_dungeon_items() + self.itempool
-
+    def get_itempool_with_dungeon_items(self) -> Iterator[Item]:
+        yield from self.get_restricted_dungeon_items()
+        yield from self.itempool
```

Comment:
IIs this change purely stylistic or is there an advantage of doing it this way?

Reply:
The previous ones combines two lists and returns a new list, whereas `yield from` doesn't create a new list, just allows iteration through the existing lists. Since we only iterate over the result of that function call, there's no reason to always make an intermediary list.

Reply:
Fair enough